### ISSUE TITLE
Big Bird Sparse Attention/Sparse Softmax

### ISSUE THREAD

**Issue by afiaka87** [lucidrains/DALLE-pytorch#292]
@lucidrains CogView describes their custom sparse attention as "similar to Big Bird". There's an option to use Big Bird from DeepSpeed. But; big bird was designed for language modeling and I honestly don't know much about it or if it's even properly causal/bidirectional. I think the change would be something like this?

```python
class BigBirdSparseAttention(Attention):
    def __init__(
        self,
        *args,
        block_size = 16,
        text_seq_len = 256,
        num_random_blocks = None,
        num_global_blocks = None,

        **kwargs
    ):
        super().__init__(*args, **kwargs)
        from deepspeed.ops.sparse_attention import SparseSelfAttention, BigBirdSparsityConfig
        self.block_size = block_size

        num_random_blocks = default(num_random_blocks, self.seq_len // block_size // 4)
        #global_block_indices = list(range(ceil(text_seq_len / block_size)))

        self.attn_fn = SparseSelfAttention(
            sparsity_config = BigBirdSparsityConfig(
                num_heads = self.heads,
                block = self.block_size,
                num_random_blocks = num_random_blocks,
                num_global_blocks = num_global_blocks,
                #global_block_indices = global_block_indices,
                #attention = 'unidirectional' if self.causal else 'bidirectional'
            ),
            max_seq_length = self.seq_len,
            attn_mask_mode = 'add'
```

There's also an efficient Softmax implementation provided which I'm curious to try out. Don't have available compute at the moment but I'll look into trying both of these.
